ios/engine/KMEI/KeymanEngine/Classes/SubKeysView.swift

```diff
@@ -0,0 +1,258 @@
+
+    let cx: CGFloat = 1.5
+    let containerWidth = (columns < 2 ? cx : CGFloat(columns)) * (buttonWidth + marginX) + marginX
+    let containerHeight = (CGFloat(rows) * (buttonHeight + marginY)) + marginY
+    let containerFrame = CGRect(x: strokeWidth, y: strokeWidth, width: containerWidth, height: containerHeight)
+    let viewWidth = containerWidth + 2 * strokeWidth
+    let baseHeight = keyFrame.size.height + marginY
+    var viewHeight = baseHeight + containerHeight - marginY + strokeWidth
+    var viewPosX = keyFrame.origin.x - (viewWidth - keyFrame.size.width) / 2.0
+    var viewPosY = keyFrame.origin.y - (viewHeight - keyFrame.size.height)
+    adjX = 0
+    adjY = 0
+    if isSystemKeyboard && (viewPosY < -tbHeight) {
+      adjY = viewPosY + tbHeight
+      viewPosY = -tbHeight
+      viewHeight += adjY
+    }
+
+    if viewPosX < 0 {
+      if (keyFrame.origin.x - borderRadius * 1.25) < 0 {
+        adjX = keyFrame.origin.x - viewPosX
+        viewPosX = keyFrame.origin.x
+      } else {
+        adjX = -viewPosX
+        viewPosX = 0
+      }
+    } else if (viewPosX + viewWidth) > screenWidth {
+      if ((keyFrame.origin.x + keyFrame.size.width) + borderRadius * 1.25) > screenWidth {
+        adjX = viewPosX - keyFrame.origin.x
+        viewPosX = keyFrame.origin.x + 2 * adjX
+      } else {
+        adjX = (screenWidth - viewWidth) - viewPosX
+        viewPosX = adjX + viewPosX
+      }
+    }
+    containerView = UIView(frame: containerFrame)
+    containerView.backgroundColor = UIColor.clear
+
+    super.init(frame: CGRect(x: viewPosX, y: viewPosY, width: viewWidth, height: viewHeight))
+
+    super.backgroundColor = UIColor.clear
+    isUserInteractionEnabled = false
+
+    addSubview(containerView)
+    let fontSize = buttonHeight * (UIDevice.current.userInterfaceIdiom == .pad ? 0.4 : 0.5)
+
+    for button in subKeys {
+      let i = button.tag
+      let buttonRow = CGFloat(i / columns)
+      let buttonFrame = CGRect(x: marginX + CGFloat(i % columns) * (buttonWidth + marginX),
+                               y: (buttonRow * buttonHeight) + (buttonRow + 1) * marginY,
+                               width: buttonWidth, height: buttonHeight)
+      button.frame = buttonFrame
+      if subKeys.count == 1 {
+        button.center = containerView.center
+      }
+
+      button.titleLabel?.minimumScaleFactor = 0.25
+      button.titleLabel?.adjustsFontSizeToFitWidth = true
+      button.titleLabel?.font = button.titleLabel?.font.withSize(fontSize)
+      containerView.addSubview(button)
+    }
+
+  }
+
+  public required init?(coder aDecoder: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
+
+  public override func draw(_ rect: CGRect) {
+    drawKeyPreviewUp(rect)
+  }
+
+  private func drawKeyPreviewUp(_ rect: CGRect) {
+    let keyWidth = keyFrame.size.width
+    let keyHeight = keyFrame.size.height + adjY
+    let viewWidth = rect.size.width
```

Comment:
There was an [additional 'forPad'](https://github.com/keymanapp/keyman/pull/378/files#diff-493470ef0a56485576a221efc4fa7c82L338) version of this method - I assume that the two were effectively identical, hence the lack of differentiation in the new Swift version?

Reply:
The call to the forPad version was commented out so I just removed it.